src/content/post/stacksafe-taming-recursive-algorithms-in-rust-without-stack-overflow.mdx

```diff
@@ -0,0 +1,272 @@
+
+1. The algorithm transforms data structures rather than just evaluating them (e.g., optimizing an AST)
+2. Multiple recursive calls need to be coordinated (e.g., tree balancing algorithms)
+3. The algorithm doesn't fit the tail-recursion pattern
+
+### Lower-Level Crates: `stacker` and `recursive`
+
+- [stacker](https://crates.io/crates/stacker): Provides low-level stack growth mechanisms
+- [recursive](https://crates.io/crates/recursive): Provides a `#[recursive]` to ease the application of `stacker`
+
+**Limitations**:
+
+- You must carefully not leaving any recursive functions not annotated with `#[recursive]`
+- Derived traits like `Debug`, `Clone`, and `Drop` on deeply nested structures still cause stack overflow, you must manually implement all traits with stack protection:
+
+```rust
+#[derive(Clone, Debug)]  // These will still stack overflow!
+enum Tree {
+    Leaf(i32),
+    Node(Box<Tree>, Box<Tree>),
+}
+
+let deep_tree = create_deep_tree(10000);
+let cloned = deep_tree.clone();  // Stack overflow here
+println!("{:?}", cloned);        // Or here
+```
+
+## `StackSafe`: The Complete Solution
+
+`StackSafe` addresses both recursive functions and recursive data structures with a simple, unified approach.
+
+### Recursive Functions Made Safe
+
+Transform any recursive function by adding `#[stacksafe]`:
+
+```rust
+use stacksafe::stacksafe;
+
+#[stacksafe]
+fn fibonacci(n: u64) -> u64 {
+    match n {
+        0 | 1 => n,
+        _ => fibonacci(n - 1) + fibonacci(n - 2),
+    }
+}
+
+#[stacksafe]
+fn evaluate_ast(expr: &Expr) -> i32 {
+    match expr {
+        Expr::Number(n) => *n,
+        Expr::Add(left, right) => evaluate_ast(left) + evaluate_ast(right),
+        Expr::Multiply(left, right) => evaluate_ast(left) * evaluate_ast(right),
+    }
+}
+```
+
+### Recursive Data Structures Made Safe
+
+Wrap recursive fields with `StackSafe<T>` for automatic stack-safe trait implementations:
+
+```rust
+use stacksafe::{stacksafe, StackSafe};
+
+#[derive(Debug, Clone, PartialEq)]  // All traits are automatically stack-safe!
+enum BinaryTree {
+    Leaf(i32),
+    Node {
+        value: i32,
+        left: StackSafe<Box<BinaryTree>>,
+        right: StackSafe<Box<BinaryTree>>,
+    },
+}
+
+// All operations work safely on arbitrarily deep trees:
+let deep_tree = create_deep_tree(100000);
+let cloned = deep_tree.clone();           // No stack overflow
+let are_equal = deep_tree == cloned;      // No stack overflow
+println!("{:?}", deep_tree);              // No stack overflow
+drop(deep_tree);                          // No stack overflow
+```
+
+### Debug-Time Safety Checks
+
+`StackSafe<T>` exposes the wrapped value through Rust's `Deref` trait, allowing transparent access to the underlying data. However, it includes an important safety mechanism: in debug builds, it checks whether the current function is properly annotated with `#[stacksafe]` whenever you access the wrapped value.
```

Comment:
This sentence reads weird: why "However"?

It reads like you must use `StackSafe<T>` to activate this debug check?